doc/developer/design/20231204_query_lifecycle_events_logging.md

```diff
@@ -0,0 +1,149 @@
+
+## The Problem
+
+Materialize is unable to communicate to users why it takes a long time
+for their queries to return, which can roughly be broken down into: 
+
+* Time for all transitive source (in the storage sense) dependencies to reach the selected
+  timestamp
+* Time for compute dependencies to reach that timestamp
+* Time for the query to finish executing (this includes the top-level
+  computation as well as returning its results to `environmentd`)
+* Time for us to flush all results to the network.
+
+With the activity log feature, users can now see how long it takes for
+execution to finish (corresponding to the 3rd point above). However,
+this has a few flaws:
+
+* It is not granular enough: in general users (or field engineers)
+need to know _why_ queries are slow in order to fix them. Is a compute
+cluster rebooting? Is a storage cluster the bottleneck?
+
+* We do not guarantee 100% sampling, so it's unreliable for debugging.
+
+
+
+## Success Criteria
+
+Users should have a reliable way to detect slow queries and
+attribute them conceptually to one of the parts of the system as
+described above.
+
+
+## Out of Scope
+
+### "Synchronous" sources
+
+Frank's [initial
+doc](https://www.notion.so/materialize/A-Model-for-Materialize-s-Operation-831f0a35782547518a8c05617dc087ca?showMoveTo=true&saveParent=true)
+describes how Storage sources should be made to work in an ideal
+world: after Materialize advances to a given timestamp, we should take
+note of the amount of data present in the upstream system
+corresponding to the source (for example, the max offset in Kafka),
+and only advance the source to that timestamp when all that data has
+been ingested. This is one of the parts of the currently
+semi-abandoned project called "real-time recency".
+
+I am leaving this out of scope for now as it would massively blow up
+the size of the project (from adding some relatively simple logging to
+conceptually re-architecting how Storage works). Unfortunately, this
+means that the time at which Storage sources reached a particular timestamp
+will not be very meaningful, as Storage is always allowed to just
+record things at however advanced of a timestamp it wants.
+
+However, we should still track and log how long it takes sources to
+update to the chosen timestamp, in the hope that someday we will do
+the work to make storage work this way. 
+
+### Time to result delivery
+
+I originally attempted to report the time it took to deliver the
+results over the network. However, after some thought, I realized
+there's no good way to do this. We could try to detect when all the
+bytes we've been sent have been `ACK`ed by repeatedly polling a
+Linux system call to get this information. However, this would not
+measure just the time to send the data, but also the tiem for the ACK
+to be returned to us.
+
+## Solution Proposal
+
+Create a collection called `mz_internal.mz_query_lifecycle_events`
```

Comment:
What's the argument for/against making these fields in the existing `mz_statement_execution_history` table?

Also: "query" not "statement"?

Reply:
Thinking about this more: I feel like we should allow for the possibility of DDL events in this table in the future. They'd just emit a somewhat different set of events. But e.g. `execution_began_at` would happen for both.

Reply:
Actually, the implementation PR I have now does exactly that. So I'll change the name to something that doesn't include `query`.